src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs

```diff
@@ -122,7 +129,20 @@ public int KeyExchangeStrength
 
     private async Task<X509Certificate2?> GetClientCertificateAsyncCore(CancellationToken cancellationToken)
     {
-        await _sslStream.NegotiateClientCertificateAsync(cancellationToken);
+        try
+        {
+            await _sslStream.NegotiateClientCertificateAsync(cancellationToken);
+        }
+        catch
+        {
+            // We can't tell which exceptions are fatal or recoverable. Consider them all fatal
+            // and close the connection to avoid over-caching and affecting future requests on this connection.
+            // This allows recovery by starting a new connection. The close is graceful to allow the server to
+            // send an error response like 401. https://github.com/dotnet/aspnetcore/issues/41369
+            _context.Features.Get<IConnectionLifetimeNotificationFeature>()?.RequestClose();
```

Comment:
Nit: It's not really fatal if you can still write a response. We should log something if IConnectionLifetimeNotificationFeature is somehow missing.

```suggestion
            // We can't tell which exceptions are fatal or recoverable. Consider them all recoverable only given a new connection
            // and close the connection gracefully to avoid over-caching and affecting future requests on this connection.
            // This allows recovery by starting a new connection. The close is graceful to allow the server to
            // send an error response like 401. https://github.com/dotnet/aspnetcore/issues/41369
            _context.Features.Get<IConnectionLifetimeNotificationFeature>()?.RequestClose();
```

Reply:
> We should log something if IConnectionLifetimeNotificationFeature is somehow missing.

That seems unlikely outside of tests? Maybe custom transports?